### ISSUE TITLE
Gradient accmulation with LSTM achieving lower accuracy on testset

### ISSUE THREAD

**Issue by hx1997** [tensorflow/tensorflow#51809]
Hi,

Sorry if this is not the right place. I'm really at a loss for my problems.
I'm using TF 2.4, and I'm trying to do gradient accumulation with my model, which is three layers of LSTM connected to three Dense layers. I'm also using a "sub-network" trick-- in each layer of my model, the first x% neurons are taken to form a sub-network, with its weights shared with the full network. (For example, if the full net is two Dense layers with 10 and 20 neurons each, then a sub-network with x being 50% is just two Dense layers with 5 and 10 neurons each.) There are three sub-networks currently, each with a different value of said x (named width_mult in the following code). In the training, I add up the gradients for the full network and the three sub-networks, and then update the weights.

The problem is, by training in this way, I'm getting lower accuracy on testset (~82% or even lower) than direct training (~89%, i.e., without gradient accumulation and the sub-network thing). But it seems that the accuracies for both full and sub-network during training are high. Also, if I get rid of the LSTM layers, the testing accuracy goes up. So I'm wondering if there is something wrong with my implementation (especially in the gradients padding part, see below), but I can't spot a bug myself. Can someone help? Perhaps a glaring error that I overlooked?

This is my custom training loop:

```python
def train_step(x, y, width_mult_list):
    train_vars = model.trainable_variables
    accum_gradient = [tf.zeros_like(this_var) for this_var in train_vars]

    with tf.GradientTape() as tape:
        # forward pass for full network
        max_predictions = model(x)
        loss = tf.keras.losses.CategoricalCrossentropy()(y, max_predictions)
    print('fullnet batch loss {}'.format(loss))

    grads = tape.gradient(loss, model.trainable_variables)

    lstm_var_cnt = 0
    for model_var in model.trainable_variables:
        print(model_var.name)
        if 'LSTM' in model_var.name: lstm_var_cnt += 1

    # accumulate gradients for full network
    accum_gradient = [(acum_grad+grad) for acum_grad, grad in zip(accum_gradient, grads)]

    sub_loss = 0

    for width_mult in sorted(width_mult_list, reverse=True):
        print(width_mult)
        # generate a sub-network with parameter width_mult
        subnet_model = utils_learning_rnn.my_model(**clf_best.filter_sk_params(utils_learning_rnn.my_model), width_mult=width_mult, fullnet_model=model)

        # we wanted to ensure the data fed into sub-network has the same number of batches as that fed into the full network,
        # so that full- and sub-network batches can match
        subnet_batch = subnet_train_dataset[fullnet_batch_num]
        subnet_features, subnet_labels = subnet_batch[0], subnet_batch[1]

        print("Start of subnet batch {}".format(batch))
        labels_c = to_categorical(subnet_labels, num_classes=num_classes)

        with tf.GradientTape() as tape:
            # forward pass for sub-network
            predictions = subnet_model(subnet_features)
            sub_loss = tf.keras.losses.CategoricalCrossentropy()(labels_c, predictions)

        grads = tape.gradient(sub_loss, subnet_model.trainable_variables)

        # LSTM gradients padding, each grad matrix is divided into 4 submatrices, corresponding to input, forget, cell, and output gate
        # we need to pad the 4 submatrices respectively and then put them back together into a big weight matrix
        # zero padding is needed because full network and sub-network are different in size, so their gradients have different shapes
        for i in range(lstm_var_cnt):
            accum_gradient_np = accum_gradient[i].numpy()
            grads_np = grads[i].numpy()
            if grads_np.ndim == 1:
                grads_np_i = grads_np[:grads_np.shape[0] // 4]
                grads_np_f = grads_np[grads_np.shape[0] // 4: grads_np.shape[0] // 4 * 2]
                grads_np_c = grads_np[grads_np.shape[0] // 4 * 2: grads_np.shape[0] // 4 * 3]
                grads_np_o = grads_np[grads_np.shape[0] // 4 * 3:]
                grads_np_i = np.pad(grads_np_i, (0, accum_gradient_np.shape[0]//4-grads_np.shape[0]//4), mode='constant')
                grads_np_f = np.pad(grads_np_f, (0, accum_gradient_np.shape[0]//4-grads_np.shape[0]//4), mode='constant')
                grads_np_c = np.pad(grads_np_c, (0, accum_gradient_np.shape[0]//4-grads_np.shape[0]//4), mode='constant')
                grads_np_o = np.pad(grads_np_o, (0, accum_gradient_np.shape[0]//4-grads_np.shape[0]//4), mode='constant')
                grads_np = np.hstack((grads_np_i, grads_np_f, grads_np_c, grads_np_o))
            elif grads_np.ndim == 2:
                grads_np_i = grads_np[:, :grads_np.shape[1] // 4]
                grads_np_f = grads_np[:, grads_np.shape[1] // 4: grads_np.shape[1] // 4 * 2]
                grads_np_c = grads_np[:, grads_np.shape[1] // 4 * 2: grads_np.shape[1] // 4 * 3]
                grads_np_o = grads_np[:, grads_np.shape[1] // 4 * 3:]
                grads_np_i = np.pad(grads_np_i, ((0, accum_gradient_np.shape[0]-grads_np.shape[0]), (0, accum_gradient_np.shape[1]//4-grads_np.shape[1]//4)), mode='constant')
                grads_np_f = np.pad(grads_np_f, ((0, accum_gradient_np.shape[0]-grads_np.shape[0]), (0, accum_gradient_np.shape[1]//4-grads_np.shape[1]//4)), mode='constant')
                grads_np_c = np.pad(grads_np_c, ((0, accum_gradient_np.shape[0]-grads_np.shape[0]), (0, accum_gradient_np.shape[1]//4-grads_np.shape[1]//4)), mode='constant')
                grads_np_o = np.pad(grads_np_o, ((0, accum_gradient_np.shape[0]-grads_np.shape[0]), (0, accum_gradient_np.shape[1]//4-grads_np.shape[1]//4)), mode='constant')
                grads_np = np.hstack((grads_np_i, grads_np_f, grads_np_c, grads_np_o))
            grads[i] = tf.convert_to_tensor(grads_np)

        # gradients padding for Dense layers
        for i in range(lstm_var_cnt, len(grads)):
            accum_gradient_np = accum_gradient[i].numpy()
            grads_np = grads[i].numpy()
            if grads_np.ndim == 1:
                grads_np = np.pad(grads_np, (0, accum_gradient_np.shape[0]-grads_np.shape[0]), mode='constant')
            elif grads_np.ndim == 2:
                grads_np = np.pad(grads_np, ((0, accum_gradient_np.shape[0]-grads_np.shape[0]), (0, accum_gradient_np.shape[1]-grads_np.shape[1])), mode='constant')
            grads[i] = tf.convert_to_tensor(grads_np)

        # accumulate gradients for sub-network
        accum_gradient = [(acum_grad+grad) for acum_grad, grad in zip(accum_gradient, grads)]

    # we were having memory issues from creating new sub-networks, so these lines are needed to release memory used
    K.clear_session()
    tf.reset_default_graph()
    del subnet_model
    gc.collect()

    # update weights using accumulated gradients
    optimizer.apply_gradients(zip(accum_gradient, model.trainable_variables))

    return loss
```

This is my model:

```python
def my_model(input_dim_eq, output_dim_eq, width_mult=1.0, fullnet_model=None, denseneurons=360, lstmlayersno=3, denselayersno=3, optimizer=Adam(lr=10e-4), nounits=288, l2reg=0.00001, verbose=True):
    activation = "relu"
    input_shape_lstm = (1, input_dim_eq)
    lstmlayersno -= 1
    denselayersno -= 2

    model = Sequential()

    return_seq_first = False if lstmlayersno == 0 else True

    if fullnet_model is None:
        model.add(LSTM(nounits, name="LSTM_0", return_sequences=return_seq_first,
                       input_shape=input_shape_lstm))
        if lstmlayersno >= 1:
            for l in range(lstmlayersno):
                return_seq_last = False if l == (lstmlayersno - 1) else True
                model.add(LSTM(nounits, name="LSTM_"+str(l+1), return_sequences=return_seq_last))

        if denselayersno >= 1:
            for l in range(denselayersno):
                model.add(Dense(denseneurons, name="Dense_"+str(l+1), kernel_regularizer=regularizers.l2(l2reg)))
                model.add(Activation(activation))


        model.add(Dense(round(denseneurons*0.8), name="deep_representation", kernel_regularizer=regularizers.l2(l2reg)))
        model.add(Activation(activation))
        model.add(Dense(output_dim_eq, name="before_softmax", kernel_regularizer=regularizers.l2(l2reg)))
        model.add(Activation('softmax'))
    else:
        # code for taking first x% (which is width_mult in this function) neurons for sub-networks, omitted here

    model.compile(loss=['categorical_crossentropy'], optimizer=optimizer,
              metrics=['accuracy', 'categorical_accuracy'])
    if verbose is True:
        print(model.summary())

    return model
```

Full network model summary:

> Model: "sequential"
> _________________________________________________________________
> Layer (type)                 Output Shape              Param #
> =================================================================
> LSTM_0 (LSTM)                (None, 1, 288)            1254528
> _________________________________________________________________
> LSTM_1 (LSTM)                (None, 1, 288)            664704
> _________________________________________________________________
> LSTM_2 (LSTM)                (None, 288)               664704
> _________________________________________________________________
> Dense_1 (Dense)              (None, 360)               104040
> _________________________________________________________________
> activation (Activation)      (None, 360)               0
> _________________________________________________________________
> deep_representation (Dense)  (None, 288)               103968
> _________________________________________________________________
> activation_1 (Activation)    (None, 288)               0
> _________________________________________________________________
> before_softmax (Dense)       (None, 204)               58956
> _________________________________________________________________
> activation_2 (Activation)    (None, 204)               0
> Total params: 2,850,900
> Trainable params: 2,850,900
> Non-trainable params: 0

Example of a sub-network model summary with x being ~95.7%:
> Model: "sequential_1"
> _________________________________________________________________
> Layer (type)                 Output Shape              Param #
> =================================================================
> LSTM_0 (LSTM)                (None, 1, 275)            1183600
> _________________________________________________________________
> LSTM_1 (LSTM)                (None, 1, 275)            606100
> _________________________________________________________________
> LSTM_2 (LSTM)                (None, 275)               606100
> _________________________________________________________________
> Dense_1 (Dense)              (None, 344)               94944
> _________________________________________________________________
> activation_3 (Activation)    (None, 344)               0
> _________________________________________________________________
> deep_representation (Dense)  (None, 275)               94875
> _________________________________________________________________
> activation_4 (Activation)    (None, 275)               0
> _________________________________________________________________
> before_softmax (Dense)       (None, 204)               56304
> _________________________________________________________________
> activation_5 (Activation)    (None, 204)               0
> Total params: 2,641,923
> Trainable params: 2,641,923
> Non-trainable params: 0

And lastly, this is my code for cropping the LSTM weight matrices. Since in sub-networks you decrease the number of neurons in each layer, the weight matrices will have to be resized (cropped) accordingly.

```python
def crop_lstm_weights(W, U, b, fullnet_nounits, subnet_nounits, input_dim=None):
    W_i = W[:, :subnet_nounits] if input_dim is None else W[:input_dim, :subnet_nounits]
    W_f = W[:, fullnet_nounits: fullnet_nounits+subnet_nounits] if input_dim is None else W[:input_dim, fullnet_nounits: fullnet_nounits+subnet_nounits]
    W_c = W[:, fullnet_nounits * 2: fullnet_nounits * 2+subnet_nounits] if input_dim is None else W[:input_dim, fullnet_nounits * 2: fullnet_nounits * 2+subnet_nounits]
    W_o = W[:, fullnet_nounits * 3: fullnet_nounits * 3+subnet_nounits] if input_dim is None else W[:input_dim, fullnet_nounits * 3: fullnet_nounits * 3+subnet_nounits]
    W_new = np.hstack((W_i, W_f, W_c, W_o))

    U_i = U[:subnet_nounits, :subnet_nounits]
    U_f = U[:subnet_nounits, fullnet_nounits: fullnet_nounits+subnet_nounits]
    U_c = U[:subnet_nounits, fullnet_nounits * 2: fullnet_nounits * 2+subnet_nounits]
    U_o = U[:subnet_nounits, fullnet_nounits * 3: fullnet_nounits * 3+subnet_nounits]
    U_new = np.hstack((U_i, U_f, U_c, U_o))

    b_i = b[:subnet_nounits]
    b_f = b[fullnet_nounits: fullnet_nounits+subnet_nounits]
    b_c = b[fullnet_nounits * 2: fullnet_nounits * 2+subnet_nounits]
    b_o = b[fullnet_nounits * 3: fullnet_nounits * 3+subnet_nounits]
    b_new = np.hstack((b_i, b_f, b_c, b_o))

    return (W_new, U_new, b_new)
```

I know this is a lot for anyone, but any help is appreciated, thanks!

**Comment by sanatmpa1** [comment]
@hx1997,

I see that you are using `TF 2.4`, I would suggest you can update to latest stable version i.e `2.6.0` and if the issue still persists, Please share a reproducible code snippet to expedite the trouble-shooting process. Thanks!

**Comment by hx1997** [comment]
Hi Sanat,

Thanks for replying! I'm a little busy recently, but I'll try your suggestions and share reproducible code if necessary.

**Comment by sanatmpa1** [comment]
@hx1997,

Sure, Please let me know if it works, or if the issue still persists in `2.6.0`. Thanks!

**Comment by hx1997** [comment]
Hi,

I (sadly) haven't got time to install TF 2.6.0 yet, but a minimized and reproducible version of my code can be found here (`grad_repro_train.py` is the training code, `utils_learning_rnn.py` is the model): https://gist.github.com/hx1997/ec2b0cf9f5e07dda1eba7b9a612aa7ff

Note that for some reason on my part, I'm not allowed to share the dataset I use to train the model, so I've replaced the input to the model with random numbers (as you can see in lines 170-174 in `grad_repro_train.py`).

I will try updating TF as soon as I got more time.

**Comment by hx1997** [comment]
Hi,

I found out the cause of my problem. It was because I didn't enable dropout in the model used in this code, whereas in direct training, I enabled it. Setting `dropout` when creating LSTM layer AND specifying `training=True` when doing forward pass raised the accuracy to ~88%, solving the problem.

Thank you for your time and attention, Sanat!
